core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -166,7 +166,7 @@ class SocketServer(val config: KafkaConfig, val metrics: Metrics, val time: Time
 private[kafka] abstract class AbstractServerThread(connectionQuotas: ConnectionQuotas) extends Runnable with Logging {
 
   private val startupLatch = new CountDownLatch(1)
-  private val shutdownLatch = new CountDownLatch(1)
+  private var shutdownLatch = new CountDownLatch(0)
```

Comment:
This should be volatile because we await on it from multiple threads. Also, it would be good to add a comment explaining why we start it with `0` and then change it to `1`.

Finally, it would be good to have a test. Should be as simple as starting a kafka server with the same config twice and verifying that the second instance shuts down correctly.